classes/submission/lesson.php

```diff
@@ -0,0 +1,109 @@
+     * @return string
+     */
+    public function get_handin_datetime(): string {
+        if ($this->submissiondata) {
+            return $this->get_iso8601_datetime($this->submissiondata->timesubmitted);
+        } else {
+            return "";
+        }
+    }
+
+    /**
+     * Set the module instance.
+     *
+     * @return void
+     * @throws \dml_exception
+     * @throws \moodle_exception
+     */
+    protected function set_module_instance() {
+        global $DB;
+        if (!$instance = $DB->get_record('lesson', ['id' => $this->coursemodule->instance])) {
+            throw new \moodle_exception(
+                'error:coursemodulenotfound', 'local_sitsgradepush', '', null,
+                "Instance ID: " . $this->coursemodule->instance
+            );
+        }
+
+        $this->modinstance = $instance;
+    }
+
+    /**
+     * Set submission.
+     *
+     * @return void
+     * @throws \dml_exception
+     * @throws \moodle_exception
+     */
+    protected function set_submission_data(): void {
+        global $DB;
+        if ($this->modinstance->practice) {
+            // If this is a "Practice Lesson" it does not appear in gradebook.
+            return;
+        }
+
+        // User may have multiple attempts if "Retakes allowed" is set to yes.
+        // In that case the grade may be the set to "Use maximum" or "Use mean".
+        // If it's "Use maximum" we return the details for the highest graded attempt here.
+        // If it's set to "Use mean", or retakes are set to not allowed, return the most recent attempt details held.
+
+        $params = ['lessonid' => $this->modinstance->id, 'userid' => $this->userid];
+
+        if ($this->modinstance->retake && $this->modinstance->usemaxgrade) {
+            $maxgrade = $DB->get_field_sql(
+                "SELECT MAX(grade) FROM {lesson_grades} WHERE lessonid = :lessonid AND userid = :userid",
+                $params
+            );
+            if ($maxgrade !== false) {
+                $params['maxgrade'] = $maxgrade;
+                $mostrecentmaxgradetime = $DB->get_field_sql(
+                "SELECT MAX(completed) FROM {lesson_grades}
+                      WHERE lessonid = :lessonid AND userid = :userid AND grade = :maxgrade",
+                    $params
+                );
+                if ($mostrecentmaxgradetime) {
+                    $this->submissiondata = (object)['timesubmitted' => $mostrecentmaxgradetime];
+                }
+            }
+        } else {
+            // We want the mean grade, or retakes is set to not allowed, so just use most recent submit time.
+            $mostrecentgradetime = $DB->get_field_sql(
+                "SELECT MAX(completed) FROM {lesson_grades}
+                      WHERE lessonid = :lessonid AND userid = :userid",
+                $params
+            );
+            if ($mostrecentgradetime) {
+                $this->submissiondata = (object)['timesubmitted' => $mostrecentgradetime];
+            }
```

Comment:
This check can be moved to the end of the function to save some lines.

Reply:
I actually think it's clearer as it is since the vars in the two instances are different $mostrecentmaxgradetime / $mostrecentgradetime